Data/ByteString/Builder.hs

```diff
@@ -445,9 +457,20 @@ charUtf8 :: Char -> Builder
 charUtf8 = P.primBounded P.charUtf8
 
 -- | UTF-8 encode a 'String'.
-{-# INLINE stringUtf8 #-}
+{-# INLINE [1] stringUtf8 #-} -- phased to allow P.cstring rewrite
 stringUtf8 :: String -> Builder
 stringUtf8 = P.primMapListBounded P.charUtf8
 
+{-# RULES
+"stringUtf8/unpackCStringUtf8#" forall s.
+  stringUtf8 (unpackCStringUtf8# s) = P.cstringUtf8 s
+
+"stringUtf8/unpackCString#" forall s.
+  stringUtf8 (unpackCString# s) = P.cstring s
```

Comment:
OK, then perhaps it's the case *below* that should be handled. You say GHC will turn `unpackCString# s` into `build (unpackFoldrCString# s)`, so you must match both in the same way. Then something is inconsistent, because below for the `build (unpackFoldrCString# s)` case you use `P.cstringUtf8`

So perhaps the actual fix is to use `P.cstring` in the `stringUtf8/unpackFoldrCString#` rule.